Allow setting permissions for user-defined functions

wmitros opened the pull request:
This series aims to allow users to set permissions on user-defined functions.

The implementation is based on Cassandra's documentation and should be fully compatible: https://cassandra.apache.org/doc/latest/cassandra/cql/security.html#cql-permissions

Fixes: #5572
Fixes: #10633

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4383/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4410/

wmitros commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4410/

It looked like an unrelated issue, so I restarted the test, but now that it failed again I'll try to debug it

wmitros commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4410/
> 
> It looked like an unrelated issue, so I restarted the test, but now that it failed again I'll try to debug it

The issue was in fact caused by my changes, but it was simple enough to fix

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4414/

wmitros commented:
This patch is quite big, but we need it to move UDFs out of experimental. @nyh I see you've already reviewed https://github.com/scylladb/scylladb/pull/10996, which is the majority of this patch. It was a long time ago, so I doubt you remember much from that review, but nevertheless, maybe reviewing this patch will be easier because of that

wmitros commented:
Sorry for pinging again @nyh but maybe this got lost in your backlog. I'm also not sure who else could help with reviewing this (@eliransin ?)

nyh commented:
Before I read the code, a nitpick on the cover letter: you say "This series aims to fix all issues that were found in https://github.com/scylladb/scylladb/pull/10996". But that is a pull request, not an issue... I guess you meant to say that this series is a new version of that old PR and replaces it. We don't need such information in the "git log", so I suggest you leave it out of the cover letter (which goes into git log) and put it in a different comment. I think we should also close #10996?

wmitros commented:
> Before I read the code, a nitpick on the cover letter: you say "This series aims to fix all issues that were found in #10996". But that is a pull request, not an issue... I guess you meant to say that this series is a new version of that old PR and replaces it. We don't need such information in the "git log", so I suggest you leave it out of the cover letter (which goes into git log) and put it in a different comment.

I actually meant that I'm trying to fix the issues _with the code_ of the https://github.com/scylladb/scylladb/pull/10996 PR, but I agree that this should not go into the git log, so I'll remove it from the cover letter.

>  I think we should also close #10996?

Yes, but I wanted to first confirm that all problems that have already been found there are fixed. For that I posted answers to the remaining comments (https://github.com/scylladb/scylladb/pull/10996#discussion_r1107178168, https://github.com/scylladb/scylladb/pull/10996#discussion_r1107178420, https://github.com/scylladb/scylladb/pull/10996#discussion_r1107179217, https://github.com/scylladb/scylladb/pull/10996#discussion_r1107181840, https://github.com/scylladb/scylladb/pull/10996#discussion_r1107182418), so if they're good we can close this already. (If not we can probably also close that and continue discussion here)

nyh reviewed (commented):
In general, the series looks very good. It's very long so I left a bunch of comments - but none of it major. I probably left incorrect comments on the first patches before I understood the later patches and then forgot to go back and correct my comments so please forgive me :-)

wmitros commented on cql3/Cql.g:
```diff
@@ -1145,7 +1145,7 @@ listPermissionsStatement returns [std::unique_ptr<list_permissions_statement> st
 
 permission returns [auth::permission perm]
-    : p=(K_CREATE | K_ALTER | K_DROP | K_SELECT | K_MODIFY | K_AUTHORIZE | K_DESCRIBE)
+    : p=(K_CREATE | K_ALTER | K_DROP | K_SELECT | K_MODIFY | K_AUTHORIZE | K_DESCRIBE | K_EXECUTE)
```
I see we're handling it in `filter_applicable_permissions()`

wmitros commented on db/cql_type_parser.cc:
```diff
@@ -37,11 +37,27 @@ data_type db::cql_type_parser::parse(const sstring& keyspace, const sstring& str
+        return *native;
+    }
+    auto raw = parse_raw(str);
+    return raw->prepare_internal(keyspace, utm).get_type();
```
This was my initial approach, but because of this, I was getting an error in some seemingly unrelated test, because it used a non-existent keyspace for the `parse` call - in the current code, it short-circuited on the native types tests, but using the one-liner we got an error on `uts.get(keyspace)`.
But now that I look at it, it's probably the test that should have been fixed, and not the change - I used this approach in the rebase

wmitros commented on cql3/statements/function_statement.cc:
```diff
@@ -24,7 +24,14 @@ future<> function_statement::check_access(query_processor& qp, const service::cl
+                _name.name,
+                ::join("^", _arg_types | boost::adaptors::transformed([] (const data_type t) {
+                    return t->name();
+                })));
```
> Fourth copy of this code, I think ;-)

I extracted it to a function in the rebase



> Also, it seems to me that the commit message isn't accurate It seems you already had code to check the ALTER permission. Looking at the test below, it seems that the code already supported "ALL FUNCTIONS" correctly (so the commit message is claiming too much), it just didn't support permissions on a specific function and this is what it appears (?) you are fixing here. So the commit message should say that.

Correct, I made the commit message more precise in the rebase

wmitros commented on test/cql-pytest/test_permissions.py:
```diff
@@ -185,3 +185,74 @@ def test_udf_permissions_quoted_names(cassandra_bug, cql):
+                        check_enforced(cql, username, permission='EXECUTE', resource=resource,
+                                function=lambda: user_session.execute(f"SELECT {keyspace}.{fun}(a) FROM {table}"))
+
+                    grant(cql, 'EXECUTE', 'ALL FUNCTIONS', username)
```
The test below is checking whether the `AUTHORIZE` permission is enforced.
Apparently, even if a user has the `AUTHORIZE` permission on some function, in order to  grant other users the `EXECUTE` permission on this function, he also needs to have the `EXECUTE` permission

wmitros commented on test/cql-pytest/test_permissions.py:
```diff
@@ -185,3 +185,74 @@ def test_udf_permissions_quoted_names(cassandra_bug, cql):
+            with new_user(cql) as username:
+                with new_session(cql, username) as user_session:
+                    fun = "fun42"
+                    grant(cql, 'CREATE', 'ALL FUNCTIONS', username)
```
Ok

wmitros replied:
It's in the few lines above this comment:
https://github.com/scylladb/scylladb/blob/ef229a5d23e96f1ae5ac05d27db50a1e163291c4/test/cql-pytest/test_permissions.py#L289-L294

wmitros commented on cql3/statements/select_statement.cc:
```diff
@@ -204,10 +204,22 @@ future<> select_statement::check_access(query_processor& qp, const service::clie
+                    used_function->name().name,
+                    ::join("^", used_function->arg_types() | boost::adaptors::transformed([] (const data_type t) {
+                        return t->name();
+                    })));
```
I made it a function in the rebase

wmitros commented on cql3/statements/select_statement.cc:
```diff
@@ -204,10 +204,22 @@ future<> select_statement::check_access(query_processor& qp, const service::clie
         auto&& s = db.find_schema(keyspace(), column_family());
         auto& cf_name = s->is_view() ? s->view_info()->base_name() : column_family();
-        return state.has_column_family_access(db, keyspace(), cf_name, auth::permission::SELECT);
+        co_await state.has_column_family_access(db, keyspace(), cf_name, auth::permission::SELECT);
```
Makes sense, but it was like that before this series and it's used in quite a few places, so maybe it's not the best idea to rename it as part of this PR

wmitros commented on test/cql-pytest/test_permissions.py:
```diff
@@ -123,3 +123,65 @@ def drop_table_idempotent():
+    schema = "a int primary key"
+    with new_test_keyspace(cql, "WITH REPLICATION = { 'class': 'SimpleStrategy', 'replication_factor': 1 }") as keyspace:
+        with new_test_table(cql, keyspace, schema) as table:
+            weird_body_lua = "(i int) CALLED ON NULL INPUT RETURNS bigint LANGUAGE lua AS 'return 42;'"
```
Ok, I'm chaging it to `fun_body`

wmitros replied:
Looks like I merged it into another commit, sorry for that

wmitros commented on test/cql-pytest/test_permissions.py:
```diff
@@ -123,3 +123,65 @@ def drop_table_idempotent():
+                with new_user(cql) as username:
+                    with new_session(cql, username) as user_session:
+                        grant(cql, 'EXECUTE', f'FUNCTION {keyspace}.{weird_fun}(int)', username)
+                        grant(cql, 'SELECT', table, username)
```
Correct (although I only tested it manually using `LIST ALL PERMISSIONS`)

wmitros commented on cql3/statements/function_statement.cc:
```diff
@@ -27,6 +28,45 @@ future<> create_function_statement_base::check_access(query_processor& qp, const
+        return state.has_function_access(qp.db(), _name.keyspace, encoded_signature, auth::permission::DROP);
+    } catch (exceptions::invalid_request_exception&) {
+        if (!_if_exists) {
+            throw;
```
From what I see, at this point in code we're not checking whether the function exists any more (we did that above), and we're only working with the `resource` object.
Consequently, the `invalid_request_exception` is actually only thrown if the given keyspace is empty: https://github.com/scylladb/scylladb/blob/25cf3256743d98f1dd34ef0656ee7cc974f87249/service/client_state.cc#L105-L108 which makes no sense here because we've already confirmed that our function is in this keyspace.

I think I'll remove the throw anyway, because as you've mentioned, `check_access()` is for checking permissions, and  `invalid_request_exception` is not an authorization problem

wmitros commented on cql3/statements/function_statement.cc:
```diff
@@ -27,6 +28,45 @@ future<> create_function_statement_base::check_access(query_processor& qp, const
+            _name.name,
+            ::join("^", _arg_types | boost::adaptors::transformed([] (const data_type t) {
+                return t->name();
+            })));
```
Yes, I added it in a rebase

wmitros commented on cql3/statements/function_statement.cc:
```diff
@@ -27,6 +28,45 @@ future<> create_function_statement_base::check_access(query_processor& qp, const
+            if (++i != funcs.end()) {
+                return make_ready_future<>();
+            }
+            func = b->second;
```
We're checking here if the range returned by `functions::functions::find` is singular by comparing begin()++ with end(). If the range is not singular, the query will fail not because of lacking permissions, but because we won't know which function to drop.
I agree that we should have a test for that, I'm adding it in the rebase

wmitros commented on auth/resource.cc:
```diff
@@ -106,6 +121,39 @@ resource::resource(role_resource_t, std::string_view role) : resource(resource_k
+resource::resource(functions_resource_t, std::string_view keyspace, std::string_view function_name, std::vector<sstring> function_signature) : resource(resource_kind::functions) {
+    _parts.emplace_back(keyspace);
+    sstring encoded_signature = format("{}[{}]",
+            function_name,
```
The workflow looks like this:

1. First we get the function name and signature string from the cql parser - if some strings were quoted in the input, they arrive unquoted from the parser
2. We encode them to the "{}[{}]" format, where the name is written unchanged before the brackets, and the types `prepare`d and then written in the brackets. Even if the arguments contained some UDTs with weird names, this is no longer a problem, because after we `prepare` them, the names are stored as bytes.
3. When we try to decode them, we look for the last `[`, because even if the function name contained `[`, no argument type has it. For the same reason we can split the arguments on '^'

So there should be no problems (as the "weird" test confirms)

wmitros replied:
I'm not seeing it documented anywhere, but after checking manually, the UDTs names are basically stored as number strings after we `prepare` them. I should have just added a test for it now that UDTs work as well. Will send a follow-up

wmitros commented on auth/service.cc:
```diff
@@ -346,6 +346,18 @@ future<bool> service::exists(const resource& r) const {
+            const auto keyspace = v.keyspace();
+            const auto function_signature = v.function_signature();
+
+            //FIXME(sarna): look up the function (if present), and signature too (if present)
```
Currently, when granting a permission on a function resource, we do not actually check if the function exists. I should have fixed this before, but I'm doing it only in the rebase.
I'm adding tests for it in the `cql-pytest: add tests for function permissions` commit

wmitros commented on auth/resource.cc:
```diff
@@ -106,6 +121,39 @@ resource::resource(role_resource_t, std::string_view role) : resource(resource_k
+    std::vector<sstring> args_types;
+    for (auto& raw_type : function_signature) {
+        // FIXME(sarna): provide information on user-defined types - this is tricky, because this information
+        // is kept in a database instance, and is thus dynamic
```
The problem was that in this context we did not know which UDTs were actually defined in the database, and we need that for the following `prepare_internal`.
I left this in the state from the original PR and fixed it in the new commits

wmitros commented:
I addressed all comments in the rebase. Additionally, I found that we didn't handle permissions on UDFs with no arguments correctly, so I fixed that and added a test for it (as part of the last commit)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4859/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4863/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4866/

nyh reviewed (approved):
Thanks. Looks good. There are a couple of places where you said you agreed with my review comment but I didn't see the fix, but maybe I just missed it? Anyway, I'll merge this PR as-is now, and if you have something additional you want to fix, please send a followup.
